modules/exploits/multi/http/tomcat_mgr_upload.rb

```diff
@@ -400,11 +413,12 @@ def undeploy_app
       return false
     end
 
-    if res and (res.code < 200 or res.code >= 300)
+    if res and (res.code < 200 or res.code >= 300 or res.body =~ /FAIL - Invalid context path/)
```

Comment:
```suggestion
    if res and (res.code < 200 or res.code >= 300 or res.body.include?('FAIL -')
```

Not saying the above code would work, but would _something_ like it that catches more one failure reason be possible?

Reply:
yeah think it's safe to just check for 'FAIL -'

from the latest Tomcat source code:
https://github.com/apache/tomcat/blob/8e2aa5e45ce13388da62386e3cb1dbfa3b242b4b/java/org/apache/catalina/manager/ManagerServlet.java#L1466

And the possible error messages in English:
https://github.com/apache/tomcat/blob/main/java/org/apache/catalina/manager/LocalStrings.properties

If undeploy is successful it outputs:
OK - Undeployed application at context path {path}

It can fail with these reasons:
FAIL - No context exists named {path}
FAIL - Context {path} is defined in server.xml and may not be undeployed
FAIL - Unable to delete {path}. The continued presence of this file may cause problems.
FAIL - The application {path} is already being serviced
FAIL - Encountered exception {path}